File: MATLAB/estimators/est_donutLSQ1_2D.m
function xest=est_donutLSQ1_2D(photonsi,patternpos,L,fwhm,background)
if nargin<5
    background=0;
end
photonsi=photonsi-background/length(photonsi);
    pi=photonsi/sum(photonsi);
    % eq 2.63
    xest=-1/(1-(L^2*log(2)/fwhm^2))*sum(pi.*patternpos);
end

File: python/estimators/est_donutLSQ1_2D.py
import numpy as np

def est_donutLSQ1_2D(photonsi, patternpos, L, fwhm, background=0):
    maxrange = L
    photonsi = photonsi - background / len(photonsi)
    
    pi = photonsi / np.sum(photonsi, axis=0)
    
    # Equation 2.63
    xest = -1 / (1 - (L**2 * np.log(2) / fwhm**2)) * np.sum(pi * patternpos,axis=0)
    xest = np.minimum(xest, maxrange)
    
    return xest